File: polygon/src/main/java/com/xattacker/android/view/polygon/PolygonMap.java
package com.xattacker.android.view.polygon;

import java.util.ArrayList;

import android.content.Context;
import android.graphics.PointF;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.annotations.SerializedName;

import com.xattacker.json.JsonBuilderVisitor;
import com.xattacker.json.JsonUtility;

public final class PolygonMap
{
	@SerializedName("width")
	private int _width;
	
	@SerializedName("height")
	private int _height;
	
	@SerializedName("border_color")
	private RegionColor _borderColor;
	
	@SerializedName("highlight_color")
	private RegionColor _highlightColor;
	
	@SerializedName("highlight_mark_color")
	private RegionColor _highlightMarkColor;
	
	@SerializedName("title_color")
	private RegionColor _titleColor;
	
	@SerializedName("regions")
	private ArrayList<PolygonRegion> _regions;
	
	public static PolygonMap parseFromJson(Context aContext, int aRes) throws Exception
	{
		PolygonMap map = JsonUtility.fromJsonRes
                       (
               		  aContext,  
               		  aRes,
                       PolygonMap.class,
                       createBuilderAdapter()
                       );
		
		return map;
	}
	
	private static JsonBuilderVisitor createBuilderAdapter()
	{
		return new JsonBuilderVisitor()
			    {
					@Override
					public void onBuilderPrepared(GsonBuilder aBuilder)
					{
						aBuilder.registerTypeAdapter(PointF.class, new PointFParserSerializer());
						aBuilder.registerTypeAdapter(RegionColor.class, new RegionColorParserSerializer());
					}
			    };
	}
	
	public PolygonMap()
	{
		_borderColor = new RegionColor();
		_highlightColor = new RegionColor();
		_titleColor = new RegionColor();
		
		_regions = new ArrayList<PolygonRegion>();
	}

	public int getWidth()
	{
		return _width;
	}

	public void setWidth(int aWidth)
	{
		_width = aWidth;
	}

	public int getHeight()
	{
		return _height;
	}

	public void setHeight(int aHeight)
	{
		_height = aHeight;
	}

	public int getBorderColor()
	{
		return _borderColor._color;
	}

	public void setBorderColor(int aBorderColor)
	{
		_borderColor._color = aBorderColor;
	}

	public int getHighlightColor()
	{
		return _highlightColor._color;
	}

	public void setHighlightColor(int aHighlightColor)
	{
		_highlightColor._color = aHighlightColor;
	}

	public void setHighlightMarkColor(int aHighlightColor)
	{
		_highlightMarkColor._color = aHighlightColor;
	}
	
	public int getHighlightMarkColor()
	{
		return _highlightMarkColor._color;
	}

	public int getTitleColor()
	{
		return _titleColor._color;
	}

	public void setTitleColor(int aTitleColor)
	{
		_titleColor._color = aTitleColor;
	}

	public ArrayList<PolygonRegion> getRegions()
	{
		return _regions;
	}
	
	public String toJson()
	{
		Gson gson = JsonUtility.createGson(createBuilderAdapter());
		return gson.toJson(this);
	}
}

File: polygon/src/main/java/com/xattacker/android/view/polygon/PolygonMap.kt
package com.xattacker.android.view.polygon

import android.content.Context
import android.graphics.Color
import android.graphics.PointF
import com.google.gson.GsonBuilder
import com.google.gson.annotations.SerializedName
import com.xattacker.json.JsonUtility
import java.util.*

class PolygonMap
{
    @SerializedName("width")
    var width: Int = 0

    @SerializedName("height")
    var height: Int = 0

    @SerializedName("border_color")
    private val _borderColor: RegionColor

    @SerializedName("highlight_color")
    private val _highlightColor: RegionColor

    @SerializedName("highlight_mark_color")
    private val _highlightMarkColor: RegionColor? = null

    @SerializedName("title_color")
    private val _titleColor: RegionColor

    @SerializedName("regions")
    val regions: ArrayList<PolygonRegion>

    var borderColor: Int
        get() = _borderColor.color
        set(aBorderColor) {
            _borderColor.color = aBorderColor
        }

    var highlightColor: Int
        get() = _highlightColor.color
        set(aHighlightColor)
        {
            _highlightColor.color = aHighlightColor
        }

    var highlightMarkColor: Int
        get() = _highlightMarkColor?.color ?: Color.TRANSPARENT
        set(aHighlightColor) {
            _highlightMarkColor?.color = aHighlightColor
        }

    var titleColor: Int
        get() = _titleColor.color
        set(aTitleColor) {
            _titleColor.color = aTitleColor
        }

    init
    {
        _borderColor = RegionColor()
        _highlightColor = RegionColor()
        _titleColor = RegionColor()

        regions = ArrayList()
    }

    fun toJson(): String {
        val gson = JsonUtility.createGson(createBuilderVisitor())
        return gson.toJson(this)
    }

    companion object
    {
        @Throws(Exception::class)
        fun parseFromJson(context: Context, aRes: Int): PolygonMap?
        {
            return JsonUtility.fromJsonRes<PolygonMap>(
                    context,
                    aRes,
                    PolygonMap::class.java,
                    createBuilderVisitor()
                    )
        }

        private fun createBuilderVisitor(): (builder: GsonBuilder) -> Unit
        {
            val visitor: (builder: GsonBuilder) -> Unit = {
                                builder: GsonBuilder ->
                                builder.registerTypeAdapter(PointF::class.java, PointFParserSerializer())
                                builder.registerTypeAdapter(RegionColor::class.java, RegionColorParserSerializer())
                            }

            return visitor
        }
    }
}
